### DISCUSSION TITLE
How to set up out-of-source build with volatile source directory name?

### DISCUSSION THREAD

**Discussion by rozhuk-im** [ccache/ccache#785] [Questions & answers] [answered]
### How to reproduce ###

1. Clone any app to dir1
2. Do out of source build 
3. Clone same app to dir2
4. Do out of source build
5. Check cache hit ratio

### Actual behavior ###

FreeBSD port system may use ccache, but cache hit ratio is very low if app change version or hash commit.

Build structure: in dir category/portname/work
- .build - here actual build happen
- .ccache - symlink to ccache dir
- portname[hashtag][version] - dir with source
- stage - dir for install

I try set CCACHE_BASEDIR=category/portname/work/portname[hashtag][version] but this not help.
Now cache has zero hit ratio if [hashtag][version] changed, if not changed - different per port, but for most, IMHO, > 80%.
My attempt to fix: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=246245

I want to do something to increase cache hit ratio then build updated port, but do not know what.

### Expected behavior ###

I expect that there is some option in ccache to allow cache hit during out of source build and source dir randomized.
Many projects use out of source build, this is not FreeBSD ports specific use case.

### Environment ###

ccache version 4.1, 3.7.1

FreeBSD 12.2 stable amd64, clang 10.0.1

**Comment by rozhuk-im** [comment]
Probably relevant issues:
https://github.com/ccache/ccache/issues/178
https://github.com/ccache/ccache/issues/485
https://github.com/ccache/ccache/issues/559
https://github.com/ccache/ccache/issues/597
https://github.com/ccache/ccache/issues/600

**Comment by AlexanderLanin** [comment]
Hi,
you have to set CCACHE_BASEDIR to `/home` or to `category` or it's parent. Something that includes all your files, but not the standard include directories. See https://ccache.dev/manual/4.1.html#config_base_dir

If you are building with `-g` have a look at #748, or even better try building release builds without `-g`.

Maybe you could provide some trivial example? e.g. does the same happen when you build ccache itself as described? or a single file?

**Comment by rozhuk-im** [comment]
I already try: ```CCACHE_BASEDIR=category/portname/work``` - this is does not work.
Probably because ```CWD=category/portname/work/.build```, so even if -I rewritten it will be ```-I../portname[hashtag][version]/src/...etc```.

I already try use ```-fdebug-prefix-map=category/portname/work/portname[hashtag][version]=.```
this works, I mean programs contain files path like src/main.c, witch should normally cache, but this does not help with hit ratio on ```[hashtag][version]``` change.

Most ports build without -g, ssdpd that was used in last tests build release, without -g.

**Comment by AlexanderLanin** [comment]
Sorry, I'm struggling a little to understand your setup.
You have a project, let's call it `coffee`.

You have two _almost_ identical source directories, like these:
`/path/to/category/coffee/work/coffee[hashtag][version]/`
`/path/to/category/coffee/work/coffee[hashtag][version]/`

You have a single build directory.
`/path/to/category/coffee/work/.build/`

When building the second version, you expect >80% hit rates on the second build, but you are getting 0%.

**Comment by rozhuk-im** [comment]
Yes.

**Comment by AlexanderLanin** [comment]
### Reproduction:
Ok first of all I can confirm that there is no hit in such a case. Here is my reproduction: https://github.com/ccache/ccache/compare/76eb7f4...AlexanderLanin:751?expand=0

Testdir result for further analysis:
https://github.com/AlexanderLanin/ccache/suites/1741089199/artifacts/33043946

### Analysis:
The actual code is identical, however lots of meta-information is different between those two builds.
Trouble is relative paths are relative to build directory and not to source directory.
Preprocessor creates two different outputs.

Example:
```
# 1 "
../dir1/include/test.h
"
int test;
```

```
# 1 "
../dir2/include/test.h
"
int test;
```

### Next steps:
Unfortunately I do not see a way to work around this issue.
Since preprocessor generates different output, how can ccache cache them as one?
IMHO this is not related to base_dir, hash_dir or debug-prefix-map at all.
Further feedback by @jrosdahl required.
This does seem like quite a reasonable use case.

#### Possible workarounds until we have a better idea:
* Compile in `/path/to/category/coffee/work/coffee[hashtag][version]/.build`
* Change directory structure so that `../src/` is always the same
`/path/to/category/coffee/work/coffee[hashtag][version]/src/`
`/path/to/category/coffee/work/coffee[hashtag][version]/.build/`
`/path/to/category/coffee/work/coffee[hashtag][version]/src/`
`/path/to/category/coffee/work/coffee[hashtag][version]/.build/`

**Comment by jrosdahl** [comment, marked as answer]
Thanks @AlexanderLanin for the analysis.

The solution to this is to make sure that the resulting relative paths don't contain volatile information, such as the mentioned hashtag and version strings.

Say that your scenario looks like this:

```
cd /path/to/category/coffee/work/.build
CCACHE_BASEDIR=/path/to ccache gcc -c /path/to/category/coffee/work/coffee_1/example.c -o /path/to/category/coffee/work/.build/example.o
# Actually executed: gcc -c ../coffee_1/example.c -o example.o
# Note: coffee_1 in pathname above, so cache miss when it's replaced with coffee_2
```

One way of fixing it is to execute the compiler from inside the source directory instead:

```
cd /path/to/category/coffee/work/coffee_1
CCACHE_BASEDIR=/path/to ccache gcc -c /path/to/category/coffee/work/coffee_1/example.c -o /path/to/category/coffee/work/.build/example.o
# Actually executed: gcc -c example.c -o ../.build/example.o
# Note: No coffee_1 in pathname above, so no cache miss when replaced with coffee_2
```

But I guess you don't have that kind of control over the build system. If so, it could be easier to use a symlink (enabled by e8444a3893cb339d97e25fd70496e448b3da18bb, included in ccache 4.0) inside the build directory like this:

```
cd /path/to/category/coffee/work/.build
ln -sf /path/to/category/coffee/work/coffee_1 src
CCACHE_BASEDIR=/path/to ccache gcc -c /path/to/category/coffee/work/.build/src/example.c -o /path/to/category/coffee/work/.build/example.o
# Actually executed: gcc -c src/example.c -o example.o
# Note: No coffee_1 in pathname above, so no cache miss when it's replaced with coffee_2
```

In other words, you need to update the symlink each time the version changes for this to work out and instruct the build to use sources from `.build/src` instead of from the real location.
